plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/serviceregistry/SystemTracker.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     * Informs all registered listeners of which systems have been added or
+     * removed from the service registry since the last refresh.
+     *
+     * @param oldSystems Systems that were previously present in the Service
+     *                   Registry.
+     * @param newSystems Systems that are present in the Service Registry.
+     */
+    private void notifyListeners(final List<? extends SrSystem> oldSystems, final List<? extends SrSystem> newSystems) {
+        // Report removed systems
+        for (final SrSystem oldSystem : oldSystems) {
+            final boolean stillPresent = newSystems.stream()
+                .anyMatch(newSystem -> newSystem.systemName().equals(oldSystem.systemName()));
+            if (!stillPresent) {
+                for (final SystemUpdateListener listener : listeners) {
+                    logger.info("System '" + oldSystem.systemName() + "' has been removed from the Service Registry.");
+                    listener.onSystemRemoved(oldSystem);
+                }
+            }
+        }
+
+        // Report added systems
+        for (final SrSystem newSystem : newSystems) {
+            final boolean wasPresent = oldSystems.stream()
+                .anyMatch(oldSystem -> newSystem.systemName().equals(oldSystem.systemName()));
+            if (!wasPresent) {
+                logger.info("System '" + newSystem.systemName() + "' detected in Service Registry.");
+                for (final SystemUpdateListener listener : listeners) {
+                    listener.onSystemAdded(newSystem);
+                }
+            }
+        }
+    }
+
+    /**
+     * Registers another object to be notified whenever a system is added or
+     * removed.
+     *
+     * @param listener Object to notify.
+     */
+    public void addListener(final SystemUpdateListener listener) {
+        Objects.requireNonNull(listener, "Expected listener.");
+        listeners.add(listener);
+    }
+
+    /**
+     * Retrieves the specified system. Note that the returned data will be stale
+     * if the system in question has changed state since the last call to {@link
+     * #fetchSystems()}. If no match is found, null is returned.
+     *
+     * @param systemName Name of a system.
+     * @param metadata   Metadata describing a system.
+     * @return The desired system, if it is present in the local cache.
+     */
+
+    public SrSystem getSystem(final String systemName, final Map<String, String> metadata) {
+        Objects.requireNonNull(systemName, "Expected system name.");
+        Objects.requireNonNull(metadata, "Expected metadata.");
+
+        if (!initialized) {
+            throw new IllegalStateException("SystemTracker has not been initialized.");
+        }
+
+        // TODO: Throw an Exception if more than one match is found?
+        for (final SrSystem system : systems) {
+            if (systemName.equals(system.systemName()) && Metadata.isSubset(metadata, system.metadata())) {
```

Comment:
The system name normalization can cause issues here.

Reply:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.